test/validate/client.go

```diff
@@ -20,6 +20,8 @@ type Validator struct {
 type IValidator interface {
 	ValidateStateFile() error
 	ValidateRestartNetwork() error
+	ValidateDualStackNodeProperties() error
```

Comment:
fyi, You will need to create an empty method for windows validator too.

Reply:
Created for windows

Reply:
@rbtr The idea was to create a `Client` for either windows/linux to validate the statefile/ips etc. The clients will be implementing this interface to provide the functionality. 
I agree we don't use the OOP patterns in Go, but this is what I added as factory method for validating in Go.
(This is what i used as reference :https://refactoring.guru/design-patterns/factory-method/go/example .)

I am open to any suggestions(correct way) to make it "Go lang" way.

Reply:
That reference is bad. Throw it away.

> The clients will be implementing this interface to provide the functionality.

I think this is where you went wrong. In Go, we do not pre-define the interface for concrete types to implement...we don't need to. Any type which provides the _behavior_ that an interface describes implicitly satisfies that interface. It's kind of upside-down from traditional OOP "inheritance".

In this way, the business logic defines the interface which describes the behavior it needs. _Any_ type which provides that behavior implicitly satisfies the interface and can be passed into that business logic. 

We don't create an `IValidator` in our validator package - that is not a useful interface definition. Where we _use_ the behavior of a validator implementation, we define an interface which describes the (subset of) that behavior that we are using. And we probably just call it `validator` and leave it package-private, because we are not defining it for implementation elsewhere, we are defining it for our immediate use in the adjacent business logic.

For example, see [here in the NNC Reconciler](https://github.com/Azure/azure-container-networking/blob/04566d5fca97d81014f910371c1ab29b0c43b3c7/cns/kubecontroller/nodenetworkconfig/reconciler.go#L25-L35): this type depends on some behavior so interfaces are defined here to abstract us from the concrete implementations of that behavior. But the implementations don't know about this interface definition, and they don't need to. Because they provide the behavior we need, they satisfy the interface.

Stop thinking about it in the OOP/inheritance style of "type A implements Iface" and go the other direction: "Iface describes the behavior I need from A (or any other type)".

Reply:
Okay, let me try this approach. I think I understand it, will ping if need help. Thank @rbtr

Reply:
https://github.com/Azure/azure-container-networking/pull/2047